### PR TITLE
Don't use cacher if uninitialized

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #49684

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher.go [hunk 1]
```diff
@@ -369,6 +369,12 @@ func (c *Cacher) Get(ctx context.Context, key string, resourceVersion string, ob
 		return err
 	}
 
+	if !c.ready.check() && getRV == 0 {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher.go]: `getRV == 0` is cheaper, check that first?

Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher.go]: also, once the cache is initialized, it seems bad to continue contending on a lock to check this forever.

Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher.go]: We are always contending on the lock.  You can't check the check without holding the lock

Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher.go]: An RLock or atomic load wouldn't make multiple readers contend

Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher.go]: Will change the order.

Can we move the change to RLock to a followup PR? We already had one Lock() before so it's not that 0 vs 1 change. I can do this tomorrow in a spearate PR.

Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher.go]: @smarterclayton just wanted to make sure this actually helps us... if the gets were blocking while the lock was already held, it wouldn't

Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher.go]: Changed the order.

**Comment by smarterclayton** [comment]
I'm investigating the problem where LISTs succeed and WATCHs always fail right now to verify that this change is sufficient to fix it.  The change as is looks good for the thing we agree on.  Looking deeper.

**Comment by wojtek-t** [comment]
> I'm investigating the problem where LISTs succeed and WATCHs always fail right now to verify that this change is sufficient to fix it. The change as is looks good for the thing we agree on. Looking deeper.

Isn't it long running LIST (but succeeding) and then WATCH failing due to "out of history" error (or whatever it is exactly called)?

**Comment by smarterclayton** [comment]
Yes, that's what I'm testing.  we see both - failures to list, and failures to watch.  They're correlated, but not coupled.  Want to verify 100% that both cases are addressed

**Comment by smarterclayton** [comment]
Tested both ways - Get/GetToList/List with this change resolves the issue I was seeing (pile up of requests behind a lock) for:

1. lists failing continuously (expected)
2. watches failing for some time until they succeed (my last comment).

/lgtm

**Comment by fejta** [comment]
/lgtm cancel

Does not build

**Comment by wojtek-t** [comment]
Just fixed compile error. Readding label.

**Pull request merged**